### PR TITLE
add CoIR tasks

### PR DESCRIPTION
**Pull request by Samoed**
## Checklist

- [X] Run tests locally to make sure nothing is broken using `make test`. 
- [X] Run the formatter to format the code using `make lint`. 

### Adding datasets checklist

**Reason for dataset addition**: #1065 

- [X] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb -m {model_name} -t {task_name}` command. https://github.com/embeddings-benchmark/results/pull/12
  - [X] `codebert-base`
  - [X] `CodeBERTa-small-v1`
- [X] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [ ] If the dataset is too big (e.g. >2048 examples), considering using `self.stratified_subsampling() under dataset_transform()`
- [X] I have filled out the metadata object in the dataset file (find documentation on it [here](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#2-creating-the-metadata-object)).
- [X] Run tests locally to make sure nothing is broken using `make test`. 
- [X] Run the formatter to format the code using `make lint`. 

Now there are not much information in `descriptive_stats`. Is there some script that can generate it?

### CODE REVIEW

**Comment by Samoed** [comment]
Dataset repositories are divided by qrels and corpus (and languages). Maybe they can be added to ignore list in test that failing

File: mteb/tasks/Retrieval/code/SyntheticText2SqlRetrieval.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+
+class SyntheticText2SQLRetrieval(AbsTaskRetrieval):
+    metadata = TaskMetadata(
+        name="SyntheticText2SQL",
+        description="The dataset is a collection of natural language queries and their corresponding sql snippets. The task is to retrieve the most relevant code snippet for a given query.",
+        reference="https://huggingface.co/datasets/gretelai/synthetic_text_to_sql",
+        dataset={
+            "path": "CoIR-Retrieval/synthetic-text2sql",
+            "revision": "main",
```
Comment (KennethEnevoldsen) [inline mteb/tasks/Retrieval/code/SyntheticText2SqlRetrieval.py:74]: should be a specific commit id (general)

**Review by KennethEnevoldsen — COMMENTED** [review]
> Dataset repositories are divided by qrels and corpus (and languages). Maybe they can be added to ignore list in test that failing

Given the license, I would re-upload them using the format presented in https://github.com/embeddings-benchmark/mteb/issues/1090

Otherwise, it would be great to get the results as well (just add them to the PR, and then we can delete them afterward).

We especially want to see how the scores match up.

**Comment by Samoed** [comment]
I've added results of these tasks to results repo https://github.com/embeddings-benchmark/results/pull/12

**Comment by KennethEnevoldsen** [comment]
> So for now these datasets should be re-uploaded to match format? How can I submit a new version of datasets? Can I make it in my hf account?

You can use the mteb hf account if you wish - I can add you if you send a user id

File: mteb/tasks/Retrieval/code/AppsRetrieval.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+          title={Measuring Coding Challenge Competence With APPS},
+          author={Dan Hendrycks and Steven Basart and Saurav Kadavath and Mantas Mazeika and Akul Arora and Ethan Guo and Collin Burns and Samir Puranik and Horace He and Dawn Song and Jacob Steinhardt},
+          journal={NeurIPS},
+          year={2021}
+        }""",
+        descriptive_stats={
+            "n_samples": {
+                _EVAL_SPLIT: 1000,
+            },
+        },
```
Comment (KennethEnevoldsen) [inline mteb/tasks/Retrieval/code/AppsRetrieval.py:97]: Seems like we need to calculate descriptive stats here. (see e.g.: https://github.com/embeddings-benchmark/mteb/blob/main/mteb/tasks/Retrieval/kor/KoStrategyQA.py) 

You can use the associated function for this: https://github.com/embeddings-benchmark/mteb/blob/main/mteb/abstasks/AbsTaskRetrieval.py#L389

**Review by KennethEnevoldsen — COMMENTED** [review]
Thanks for the results. Comparing them to the [leaderboard](https://archersama.github.io/coir/) it is hard to tell if the implementation matches. Would it be possible to run the e5 model to allow for comparison with the existing benchmark?

**Comment by Samoed** [comment]
> > So for now these datasets should be re-uploaded to match format? How can I submit a new version of datasets? Can I make it in my hf account?
> 
> You can use the mteb hf account if you wish - I can add you if you send a user id

[My hf account](https://huggingface.co/Samoed). I will run e5 models (and others) later

File: mteb/tasks/Retrieval/code/AppsRetrieval.py (hunk 2 above)
Comment (KennethEnevoldsen) [reply mteb/tasks/Retrieval/code/AppsRetrieval.py:97]: We sadly haven't added it to all datasets yet, but once added, we should definitely. do that

**Comment by Samoed** [comment]
I've run these and coir tasks with `e5-base-v2`. All tasks results are the same except for CodetransoceanContest and all CodeSearchCCR. I'll try to look through to find what's wrong.
[coir_mteb_results.zip](https://github.com/user-attachments/files/16506670/coir_mteb_results.zip)
[notebook](https://www.kaggle.com/code/samoed/notebook41710787c8/notebook)

**Comment by Samoed** [comment]
In coir benchmark they are using only text of corpus without title. I've added it to model and result are more near.
[CodeTransOceanContest.json](https://github.com/user-attachments/files/16518827/CodeTransOceanContest.json)
[codetrans-contest.json](https://github.com/user-attachments/files/16518828/codetrans-contest.json)

But results of CodeSearchNetCCR still different

<details>
  <summary>Model with title</summary>

```python
import coir
from coir.data_loader import get_tasks
from coir.evaluation import COIR
from coir.models import YourCustomDEModel
import random
import numpy as np
import torch
from sentence_transformers import SentenceTransformer

seed = 42
random.seed(seed)
np.random.seed(seed)
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)

model_name = "intfloat/e5-base-v2"

def corpus_to_texts(
    corpus: list[dict[str, str]] | dict[str, list[str]] | list[str],
    sep: str = "\n",
) -> list[str]:
    if isinstance(corpus, dict):
        return [
            (corpus["title"][i] + sep + corpus["text"][i]).strip()  # type: ignore
            if "title" in corpus
            else corpus["text"][i].strip()  # type: ignore
            for i in range(len(corpus["text"]))  # type: ignore
        ]
    else:
        if isinstance(corpus[0], str):
            return corpus
        return [
            (doc["title"] + sep + doc["text"]).strip()
            if "title" in doc
            else doc["text"].strip()
            for doc in corpus
        ]


class E5Wrapper:
    """following the implementation within the Scandinavian Embedding Benchmark and the intfloat/multilingual-e5-small documentation."""

    def __init__(
        self,
        model_name: str,
        sep: str = " ",
        prompt_name: str | None = None,
        **kwargs,
    ):
        self.model_name = model_name
        self.mdl = SentenceTransformer(model_name)
        self.sep = sep

    def to(self, device: torch.device) -> None:
        self.mdl.to(device)

    def encode(  # type: ignore
        self,
        sentences: list[str],
        *,
        batch_size: int = 32,
        **kwargs,
    ):
        return self.encode_queries(sentences, batch_size=batch_size, **kwargs)

    def encode_queries(
        self,
        queries: list[str],
        batch_size: int = 32,
        prompt_name: str | None = None,
        **kwargs,
    ):
        sentences = ["query: " + sentence for sentence in queries]
        emb = self.mdl.encode(sentences, batch_size=batch_size, **kwargs)
        return emb

    def encode_corpus(
        self,
        corpus: list[dict[str, str]] | dict[str, list[str]],
        prompt_name: str | None = None,
        batch_size: int = 32,
        **kwargs,
    ):
        if "request_qid" in kwargs:
            kwargs.pop("request_qid")
        sentences = corpus_to_texts(corpus)
        sentences = ["passage: " + sentence for sentence in sentences]
        emb = self.mdl.encode(sentences, batch_size=batch_size, **kwargs)
        return emb


# Load the model
model = E5Wrapper(model_name=model_name)

# Get tasks
tasks = get_tasks(tasks=["codetrans-contest", "CodeSearchNet-ruby", "apps"])

# Initialize evaluation
evaluation = COIR(tasks=tasks, batch_size=64)

# Run evaluation
results = evaluation.run(model, output_folder=f"coir_results/{model_name}")
print(results)
```
</details>

**Comment by archersama** [comment]
> In coir benchmark they are using only text of corpus without title. I've added it to model and result are more near. [CodeTransOceanContest.json](https://github.com/user-attachments/files/16518827/CodeTransOceanContest.json) [codetrans-contest.json](https://github.com/user-attachments/files/16518828/codetrans-contest.json)
> 
> But results of CodeSearchNetCCR still different
> 
> Model with title

Thank you for your grate work. We are the authors of CoIR and have recently been working to integrate the CoIR task into the MTEB. Using the MTEB framework, we have obtained results consistent with the COIR leaderboard. We will also be making a submission soon, and you can look forward to our contribution to addressing the existing issues.

**Comment by KennethEnevoldsen** [comment]
Thanks @archersama will be looking forward to it

**Comment by Samoed** [comment]
Now results of CodeSearchNet-CCR are the same. Previously, I mistakenly run CodeSearchNet instead of CodeSearchNet-ccr in COIR. Now everything working correctly. @KennethEnevoldsen I think PR can be merged, but since COIR team working on integration I don't know what to do whis this PR.
[Coir-CodeSearchNet-ccr-ruby.json](https://github.com/user-attachments/files/16538310/CodeSearchNet-ccr-ruby.json)
[Mteb-CodeSearchNetCCRetrieval.json](https://github.com/user-attachments/files/16538316/CodeSearchNetCCRetrieval.json)

**Comment by KennethEnevoldsen** [comment]
Thanks @Samoed! We are good to merge then. 

@archersama can then overwrite this version of the data and/or update it -  however they see fit. Hopefully, it can take some work of of their shoulder.

**Pull request merged**